### ISSUE TITLE
Error executing 'postInstallation': Database tables already present

### ISSUE THREAD

**Issue by wsalazar-peakactivity** [bitnami/bitnami-docker-moodle#95]
**Description**

I've seen that other tickets have been opened around the topic but never properly resolved. I'm trying to set Moodle up to horizontally scale on Kubernetes, but each replica fails because it tries to recreate the database each time. I've tried using the 'externalDatabase' database setting and even ReadWriteMany as one of the previous issues mentions. Nothing works. 

**Steps to reproduce the issue:**

1. Install the helm cart on Kubernetes
2. Replicate the deployment

**Describe the results you received:**

Error executing 'postInstallation': Database tables already present, cli installation can not continue.

**Describe the results you expected:**

For the deployment to horizontally scale.

**Comment by marcosbc** [comment]
Hi @wsalazar-peakactivity, currently the logic is designed so the Moodle install logic is always executed if the container had not been initialized yet (i.e. the volume is initialized).

We have done similar things in [Discourse](https://github.com/bitnami/bitnami-docker-discourse) (with `DISCOURSE_SKIP_INSTALL`) and [WordPress](https://github.com/bitnami/bitnami-docker-wordpress) (with `WORDPRESS_SKIP_INSTALL`).

Basically this requires to have an already initialized database, and for the application not to modify file at installation time.

We'll take a look at this tomorrow and let you know if it is a feasible change.

**Comment by wsalazar-peakactivity** [comment]
Thanks @marcosbc. I can also help out. If you point me in the right direction, I can fork the repo, make my changes, and PR back into the codebase. I've noticed there are multiple repositories for this setup. I started to look at https://github.com/bitnami/nami but wanted to chat your organization first before forking. I'm hoping it can be a quick fix on your side. For my organization, this is a mission critical feature and seems necessary when deploying on Kubernetes.

**Comment by marcosbc** [comment]
Hi @wsalazar-peakactivity, we were able to implement this feature without issues. Check out the `MOODLE_SKIP_INSTALL` flag as documented in the README. Let us know if this works for you!

Note that this requires to have an already initialized Moodle database, after that add `MOODLE_SKIP_INSTALL` to all nodes and it should work for replicas.

**Comment by wsalazar-peakactivity** [comment]
Thank for the quick response. Which tag should I update my helm chart with? This what I have currently:

`image:
  registry: docker.io
  repository: bitnami/moodle
  tag: 3.5.2-debian-9`

I went to the what I believe is the latest tag but not seeing the skip install feature in the docker build file:

https://github.com/bitnami/bitnami-docker-moodle/blob/3.7.1-debian-9-r38/3/debian-9/Dockerfile

Thanks.

**Comment by marcosbc** [comment]
Hi @wsalazar-peakactivity, you are right. Apparently the release didn't get triggered properly.

I re-launched the release and now it is present. Hope it works for you!

**Comment by wsalazar-peakactivity** [comment]
Thank you. When I pull down 3.7.1 I get this message. Any ideas?

moodle  INFO  The installation wizard will be skipped and parameters for configuring the user account and SMTP credentials will be ignored.
Error executing 'postInstallation': PHP Warning:  require(/opt/bitnami/moodle/admin/cli/../../config.php): failed to open stream: No such file or directory in /opt/bitnami/moodle/admin/cli/upgrade.php on line 41
PHP Fatal error:  require(): Failed opening required '/opt/bitnami/moodle/admin/cli/../../config.php' (include_path='.:/opt/bitnami/php/lib/php') in /opt/bitnami/moodle/admin/cli/upgrade.php on line 41

**Comment by wsalazar-peakactivity** [comment]
This only appears when I set MOODLE_SKIP_INSTALL to 'yes'. For my own curiosity, can you point me to the Nami Moodle package repo?  I'm assuming 'bitnami/moodle/admin/cli/upgrade.php' is part of this repo. Thanks.

**Comment by alemorcuq** [comment]
Hi, @wsalazar-peakactivity.

I've pulled `bitnami/moodle:3.7.1` and I can see that file in the container:

```
$ docker run --rm bitnami/moodle:3.7.1 bash -c 'cat /opt/bitnami/moodle/admin/cli/upgrade.php'
Unable to find image 'bitnami/moodle:3.7.1' locally
3.7.1: Pulling from bitnami/moodle
Digest: sha256:4f893bba1cd7f51dd37ec98694472c6bd6a726fa576c72280704b8b9fb212873
Status: Downloaded newer image for bitnami/moodle:3.7.1

Welcome to the Bitnami moodle container
Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-moodle
Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-moodle/issues

<?php

// This file is part of Moodle - http://moodle.org/
//
// Moodle is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Moodle is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.

/**
 * This script creates config.php file and prepares database.
 *
 * This script is not intended for beginners!
 * Potential problems:
 * - su to apache account or sudo before execution
 * - not compatible with Windows platform
 *
 * @package    core
 * @subpackage cli
 * @copyright  2009 Petr Skoda (http://skodak.org)
 * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
 */

// Force OPcache reset if used, we do not want any stale caches
// when detecting if upgrade necessary or when running upgrade.
if (function_exists('opcache_reset') and !isset($_SERVER['REMOTE_ADDR'])) {
    opcache_reset();
}

define('CLI_SCRIPT', true);
define('CACHE_DISABLE_ALL', true);

require(__DIR__.'/../../config.php');
require_once($CFG->libdir.'/adminlib.php');       // various admin-only functions
require_once($CFG->libdir.'/upgradelib.php');     // general upgrade/install related functions
require_once($CFG->libdir.'/clilib.php');         // cli only functions
require_once($CFG->libdir.'/environmentlib.php');

// now get cli options
$lang = isset($SESSION->lang) ? $SESSION->lang : $CFG->lang;
list($options, $unrecognized) = cli_get_params(
    array(
        'non-interactive'   => false,
        'allow-unstable'    => false,
        'help'              => false,
        'lang'              => $lang,
        'verbose-settings'  => false
    ),
    array(
        'h' => 'help'
    )
);

if ($options['lang']) {
    $SESSION->lang = $options['lang'];
}

$interactive = empty($options['non-interactive']);

if ($unrecognized) {
    $unrecognized = implode("\n  ", $unrecognized);
    cli_error(get_string('cliunknowoption', 'admin', $unrecognized));
}

if ($options['help']) {
    $help =
"Command line Moodle upgrade.
Please note you must execute this script with the same uid as apache!

Site defaults may be changed via local/defaults.php.

Options:
--non-interactive     No interactive questions or confirmations
--allow-unstable      Upgrade even if the version is not marked as stable yet,
                      required in non-interactive mode.
--lang=CODE           Set preferred language for CLI output. Defaults to the
                      site language if not set. Defaults to 'en' if the lang
                      parameter is invalid or if the language pack is not
                      installed.
--verbose-settings    Show new settings values. By default only the name of
                      new core or plugin settings are displayed. This option
                      outputs the new values as well as the setting name.
-h, --help            Print out this help

Example:
\$sudo -u www-data /usr/bin/php admin/cli/upgrade.php
"; //TODO: localize - to be translated later when everything is finished

    echo $help;
    die;
}

if (empty($CFG->version)) {
    cli_error(get_string('missingconfigversion', 'debug'));
}

require("$CFG->dirroot/version.php");       // defines $version, $release, $branch and $maturity
$CFG->target_release = $release;            // used during installation and upgrades

if ($version < $CFG->version) {
    cli_error(get_string('downgradedcore', 'error'));
}

$oldversion = "$CFG->release ($CFG->version)";
$newversion = "$release ($version)";

if (!moodle_needs_upgrading()) {
    cli_error(get_string('cliupgradenoneed', 'core_admin', $newversion), 0);
}

// Test environment first.
list($envstatus, $environment_results) = check_moodle_environment(normalize_version($release), ENV_SELECT_RELEASE);
if (!$envstatus) {
    $errors = environment_get_errors($environment_results);
    cli_heading(get_string('environment', 'admin'));
    foreach ($errors as $error) {
        list($info, $report) = $error;
        echo "!! $info !!\n$report\n\n";
    }
    exit(1);
}

// Test plugin dependencies.
$failed = array();
if (!core_plugin_manager::instance()->all_plugins_ok($version, $failed)) {
    cli_problem(get_string('pluginscheckfailed', 'admin', array('pluginslist' => implode(', ', array_unique($failed)))));
    cli_error(get_string('pluginschecktodo', 'admin'));
}

$a = new stdClass();
$a->oldversion = $oldversion;
$a->newversion = $newversion;

if ($interactive) {
    echo cli_heading(get_string('databasechecking', '', $a)) . PHP_EOL;
}

// make sure we are upgrading to a stable release or display a warning
if (isset($maturity)) {
    if (($maturity < MATURITY_STABLE) and !$options['allow-unstable']) {
        $maturitylevel = get_string('maturity'.$maturity, 'admin');

        if ($interactive) {
            cli_separator();
            cli_heading(get_string('notice'));
            echo get_string('maturitycorewarning', 'admin', $maturitylevel) . PHP_EOL;
            echo get_string('morehelp') . ': ' . get_docs_url('admin/versions') . PHP_EOL;
            cli_separator();
        } else {
            cli_problem(get_string('maturitycorewarning', 'admin', $maturitylevel));
            cli_error(get_string('maturityallowunstable', 'admin'));
        }
    }
}

if ($interactive) {
    echo html_to_text(get_string('upgradesure', 'admin', $newversion))."\n";
    $prompt = get_string('cliyesnoprompt', 'admin');
    $input = cli_input($prompt, '', array(get_string('clianswerno', 'admin'), get_string('cliansweryes', 'admin')));
    if ($input == get_string('clianswerno', 'admin')) {
        exit(1);
    }
}

if ($version > $CFG->version) {
    // We purge all of MUC's caches here.
    // Caches are disabled for upgrade by CACHE_DISABLE_ALL so we must set the first arg to true.
    // This ensures a real config object is loaded and the stores will be purged.
    // This is the only way we can purge custom caches such as memcache or APC.
    // Note: all other calls to caches will still used the disabled API.
    cache_helper::purge_all(true);
    upgrade_core($version, true);
}
set_config('release', $release);
set_config('branch', $branch);

// unconditionally upgrade
upgrade_noncore(true);

// log in as admin - we need doanything permission when applying defaults
\core\session\manager::set_user(get_admin());

// Apply default settings and output those that have changed.
cli_heading(get_string('cliupgradedefaultheading', 'admin'));
$settingsoutput = admin_apply_default_settings(null, false);

foreach ($settingsoutput as $setting => $value) {

    if ($options['verbose-settings']) {
        $stringvlaues = array(
                'name' => $setting,
                'defaultsetting' => var_export($value, true) // Expand objects.
        );
        echo get_string('cliupgradedefaultverbose', 'admin', $stringvlaues) . PHP_EOL;

    } else {
        echo get_string('cliupgradedefault', 'admin', $setting) . PHP_EOL;

    }
}

// This needs to happen at the end to ensure it occurs after all caches
// have been purged for the last time.
// This will build a cached version of the current theme for the user
// to immediately start browsing the site.
upgrade_themes();

echo get_string('cliupgradefinished', 'admin', $a)."\n";
exit(0); // 0 means success
```

Could you check if you have that file present in your container? This file is also part of Moodle itself, you can find it here: https://github.com/moodle/moodle/blob/master/admin/cli/upgrade.php

On the other hand, the Nami package is not in a public repo, but you can find all the code inside the moodle container, under the `/root/.nami` directory.

Regards,
Alejandro

**Comment by marcosbc** [comment]
Hi @wsalazar-peakactivity, you are right about that. The config.php file is creating during the installation wizard, we're working on fixing that.

**Comment by johnysd-GR** [comment]
Hello everyone.
I 'm facing a similar issue and I'll be glad if you can help.
Thank you in advance
https://community.bitnami.com/t/migrate-bitnami-moodle-from-aws-ec2-to-aws-ecs/70179

**Comment by marcosbc** [comment]
Hi @wsalazar-peakactivity, we released Moodle 3.7.1-debian-9-r45 and verified it fixes this issue.

@johnysd-GR Please check out the `MOODLE_SKIP_INSTALL` flag. If you enable it, the database will not be re-initialized.

**Comment by wsalazar-peakactivity** [comment]
I tested it, and it works as expected. I appreciate the quick turnaround on this. Thank you.

**Comment by marcosbc** [comment]
Thanks for letting us know @wsalazar-peakactivity! We'll close this issue then.
